standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/TestObjectStore.java

```diff
@@ -1483,6 +1483,135 @@ public void testListPackage() throws Exception {
+    if (!origUseTableSerDes) {
+      MetastoreConf.setBoolVar(conf, ConfVars.USE_TABLE_SERDES, true);
+      objectStore.setConf(conf);
+    }
+
+    createPartitionedTable(true, true);
+    // Partitions should reuse table's serde info
+    checkBackendTableSize("PARTITIONS", 3);
+    checkBackendTableSize("SERDES", 1);
+    checkBackendTableSize("SERDE_PARAMS", 1);
+    Table tbl;
+    Partition newPart;
+    // Alters table's serde info
+    try (AutoCloseable c = deadline()) {
+      tbl = objectStore.getTable(DEFAULT_CATALOG_NAME, DB1, TABLE1);
+      Table newTbl = tbl.deepCopy();
+      newTbl.getSd().getSerdeInfo().setDescription("To test SerDe is created on demand");
+      objectStore.alterTable(DEFAULT_CATALOG_NAME, DB1, TABLE1, newTbl, null);
+    }
+    // A new SERDE should be created
+    checkBackendTableSize("SERDES", 2);
+    checkBackendTableSize("SERDE_PARAMS", 2);
+    // Alter a partition's serde info
+    List<String> partVals = Collections.singletonList("a0");
+    try (AutoCloseable c = deadline()) {
+      Partition part = objectStore.getPartition(DEFAULT_CATALOG_NAME, DB1, TABLE1, partVals);
+      newPart = part.deepCopy();
+      newPart.getSd().getSerdeInfo().setDescription("To test SerDe is created on demand");
+      objectStore.alterPartition(DEFAULT_CATALOG_NAME, DB1, TABLE1, partVals, newPart, null);
+    }
+    // A new SERDE should be created
+    checkBackendTableSize("SERDES", 3);
+    checkBackendTableSize("SERDE_PARAMS", 3);
+    // Adds a partition with different serde
+    newPart.setValues(Collections.singletonList("a3"));
+    newPart.getSd().getSerdeInfo().setDescription("To test adding a partition with different SerDe");
+    objectStore.addPartition(newPart);
+    // A new SERDE should be created
+    checkBackendTableSize("SERDES", 4);
+    checkBackendTableSize("SERDE_PARAMS", 4);
+
+    // Restore conf
+    if (!origUseTableSerDes) {
+      MetastoreConf.setBoolVar(conf, ConfVars.USE_TABLE_SERDES, false);
+      objectStore.setConf(conf);
+    }
+  }
+
+  @Test
+  public void testSerDesCleanup() throws Exception {
+    // Enable USE_TABLE_SERDES
+    final boolean origUseTableSerDes = MetastoreConf.getBoolVar(conf, ConfVars.USE_TABLE_SERDES);
+    if (!origUseTableSerDes) {
+      MetastoreConf.setBoolVar(conf, ConfVars.USE_TABLE_SERDES, true);
+      objectStore.setConf(conf);
+    }
+
+    createPartitionedTable(true, true);
+    // Partitions should reuse table's serde info
+    checkBackendTableSize("PARTITIONS", 3);
+    checkBackendTableSize("SERDES", 1);
+    checkBackendTableSize("SERDE_PARAMS", 1);
+    // Alters a partition's serde info
+    List<String> partVals = Collections.singletonList("a0");
+    Partition newPart;
+    try (AutoCloseable c = deadline()) {
+      Partition part = objectStore.getPartition(DEFAULT_CATALOG_NAME, DB1, TABLE1, partVals);
+      newPart = part.deepCopy();
```

Comment:
could we should check the serde info before and after the change?

Reply:
@dengzhhu653 I'm not sure if I understand it. Did you mean checking the serde info before and after `alterPartition`?

Reply:
yeah, that it.